base/linalg/rowvector.jl

```diff
@@ -0,0 +1,214 @@
+"""
+    RowVector(vector)
+
+A lazy-view wrapper of an `AbstractVector`, which turns a length-`n` vector into
+a `1×n` shaped row vector and represents the transpose of a vector (the elements
+are also transposed recursively). This type is usually constructed (and
+unwrapped) via the `transpose()` function or `.'` operator (or related
+`ctranspose()` or `'` operator).
+
+By convention, a vector can be multiplied by a matrix on its left (`A * v`)
+whereas a row vector can be multiplied by a matrix on its right (such that
+`v.' * A = (A.' * v).'`). It differs from a `1×n`-sized matrix by the facts that
+its transpose returns a vector and the inner product `v1.' * v2` returns a
+scalar, but will otherwise behave similarly.
+"""
+immutable RowVector{T,V<:AbstractVector} <: AbstractMatrix{T}
+    vec::V
+    function RowVector(v::V)
+        check_types(T,v)
+        new(v)
+    end
+end
+
+
+@inline check_types{T1,T2}(::Type{T1},::AbstractVector{T2}) = check_types(T1, T2)
+@pure check_types{T1,T2}(::Type{T1},::Type{T2}) = T1 === transpose_type(T2) ? nothing : error("Element type mismatch. Tried to create a `RowVector{$T1}` from an `AbstractVector{$T2}`")
+
+# The element type may be transformed as transpose is recursive
+@inline transpose_type{T}(::Type{T}) = promote_op(transpose, T)
```

Comment:
True. Hopefully `transpose` is an inferrable operation. Typically, people pay attention to the performance (and inferrability) of arrays.